src/script/interpreter.cpp

```diff
@@ -2051,7 +2051,36 @@ static bool VerifyTaprootSpend(std::vector<valtype> stack,
     if (!IsPayToTaproot(script_pubkey)) {
         return set_error(serror, ScriptError::SCRIPTTYPE_MALFORMED_SCRIPT);
     }
-    // Temporarily return UNKNOWN
+    valtype vch_pubkey =
+        valtype(script_pubkey.begin() + TAPROOT_INTRO_SIZE,
+                script_pubkey.begin() + TAPROOT_SIZE_WITHOUT_STATE);
+
+    if (stack.size() == 0) {
+        return set_error(serror, ScriptError::INVALID_STACK_OPERATION);
+    }
+    if (stack.size() >= 2 && !stack.back().empty() &&
+        stack.back()[0] == TAPROOT_ANNEX_TAG) {
+        return set_error(serror, ScriptError::TAPROOT_ANNEX_NOT_SUPPORTED);
+    }
+    if (stack.size() == 1) {
+        // Spend using single signature instead of executing script
+        const valtype &vch_sig = stacktop(-1);
+        const uint32_t sigflags = flags | SCRIPT_TAPROOT_KEY_SPEND_PATH;
+        if (!CheckTransactionSignatureEncoding(vch_sig, sigflags, serror) ||
+            !CheckPubKeyEncoding(vch_pubkey, sigflags, serror)) {
+            // serror is set
+            return false;
+        }
+
+        if (vch_sig.size() > 0 &&
```

Comment:
This seems fine, I'd still for verbosity and clarity move the `if(vch_sig.size() == 0) { return set_error(...}` here as well.

Reply:
```C++
if(vch_sig.size() == 0) { 
    return set_error(...}
}
const boolean fSuccess = checker.CheckSig(vch_sig, vch_pubkey, std::nullopt, {}, sigflags);
if(!fSuccess) {
   return set_error(serror,
                             ScriptError::TAPROOT_VERIFY_SIGNATURE_FAILED);
}
return set_success(serror);
```